File: packages/react-native-buffered-blob/ios/HandleRegistry.m
#import "HandleRegistry.h"

@implementation HandleRegistry {
  NSInteger _nextId;
  NSMutableDictionary<NSNumber *, id> *_handles;
  NSLock *_lock;
}

+ (instancetype)shared {
  static HandleRegistry *instance;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    instance = [[HandleRegistry alloc] initPrivate];
  });
  return instance;
}

- (instancetype)initPrivate {
  self = [super init];
  if (self) {
    _nextId = 1;
    _handles = [NSMutableDictionary new];
    _lock = [NSLock new];
  }
  return self;
}

- (NSInteger)registerObject:(id)obj {
  [_lock lock];
  NSInteger handleId = _nextId++;
  _handles[@(handleId)] = obj;
  [_lock unlock];
  return handleId;
}

- (id)objectForId:(NSInteger)handleId {
  [_lock lock];
  id obj = _handles[@(handleId)];
  [_lock unlock];
  return obj;
}

- (void)removeObjectForId:(NSInteger)handleId {
  [_lock lock];
  id obj = _handles[@(handleId)];
  [_handles removeObjectForKey:@(handleId)];
  [_lock unlock];

  if ([obj conformsToProtocol:@protocol(HandleCloseable)]) {
    [(id<HandleCloseable>)obj closeHandle];
  }
}

- (void)clear {
  [_lock lock];
  NSDictionary<NSNumber *, id> *snapshot = [_handles copy];
  [_handles removeAllObjects];
  [_lock unlock];

  for (id obj in snapshot.allValues) {
    if ([obj conformsToProtocol:@protocol(HandleCloseable)]) {
      [(id<HandleCloseable>)obj closeHandle];
    }
  }
}

@end

File: packages/react-native-buffered-blob/ios/HandleRegistry.swift
import Foundation

final class HandleRegistry {
  static let shared = HandleRegistry()

  private var nextId: Int = 1
  private var handles: [Int: AnyObject] = [:]
  private let lock = NSLock()

  private init() {}

  func register(_ obj: AnyObject) -> Int {
    lock.lock()
    defer { lock.unlock() }
    let id = nextId
    nextId += 1
    handles[id] = obj
    return id
  }

  func get<T: AnyObject>(_ id: Int) -> T? {
    lock.lock()
    defer { lock.unlock() }
    return handles[id] as? T
  }

  func remove(_ id: Int) {
    lock.lock()
    let obj = handles.removeValue(forKey: id)
    lock.unlock()
    // Close if applicable
    if let closeable = obj as? HandleCloseable {
      closeable.closeHandle()
    }
  }

  func clear() {
    lock.lock()
    let allIds = Array(handles.keys)
    lock.unlock()
    allIds.forEach { remove($0) }
  }
}

@objc protocol HandleCloseable {
  func closeHandle()
}
